File: src/dialog/SheetButton.ts
import { bridge } from '../decorator/bridge'
import { native } from '../decorator/native'
import { Event } from '../event/Event'
import { Emitter } from '../event/Emitter'
import { Image } from '../graphic/Image'

/**
 * @interface SheetButtonOptions
 * @since 0.5.0
 */
export interface SheetButtonOptions {
	id?: string
	label?: string
	style?: 'normal' | 'cancel' | 'destructive'
	image?: string | Image
}

@bridge('dezel.dialog.SheetButton')

/**
 * Displays an alert button.
 * @class SheetButton
 * @super Emitter
 * @since 0.5.0
 */
export class SheetButton extends Emitter {

	//--------------------------------------------------------------------------
	// Properties
	//--------------------------------------------------------------------------

	/**
	 * The alert's button id.
	 * @property name
	 * @since 0.4.0
	 */
	public id: string

	/**
	 * The sheet button's label.
	 * @property label
	 * @since 0.5.0
	 */
	@native public readonly label!: string

	/**
	 * The sheet button's style.
	 * @property style
	 * @since 0.5.0
	 */
	@native public readonly style!: 'normal' | 'cancel' | 'destructive'

	/**
	 * The sheet button's image.
	 * @property image
	 * @since 0.5.0
	 */
	@native public readonly image!: string | Image | null

	//--------------------------------------------------------------------------
	// Methods
	//--------------------------------------------------------------------------

	/**
	 * Initializes the alert button.
	 * @constructor
	 * @since 0.5.0
	 */
	constructor(options: SheetButtonOptions) {
		super()
		this.id = options.id || 'ok'
		this.label = options.label || 'OK'
		this.style = options.style || 'normal'
		this.image = options.image || null
	}

	//--------------------------------------------------------------------------
	// Events
	//--------------------------------------------------------------------------

	/**
	 * @inherited
	 * @method onEmit
	 * @since 0.5.0
	 */
	onEmit(event: Event) {

		switch (event.type) {

			case 'press':
				this.onPress(event)
				break
		}

		super.onEmit(event)
	}

	/**
	 * @method onPress
	 * @since 0.5.0
	 * @hidden
	 */
	onPress(event: Event) {

	}

	//--------------------------------------------------------------------------
	// Native API
	//--------------------------------------------------------------------------

	/**
	 * @property native
	 * @since 0.5.0
	 * @hidden
	 */
	public native: any

	/**
	 * @method nativePress
	 * @since 0.5.0
	 * @hidden
	 */
	private nativePress() {
		this.emit('press')
	}
}

File: platforms/ios/Dezel/Dezel/Modules/Dialog/SheetButton.swift
/**
 * @class SheetButton
 * @since 0.5.0
 * @hidden
 */
open class SheetButton: JavaScriptClass {

	//--------------------------------------------------------------------------
	// MARK: Properties
	//--------------------------------------------------------------------------

	/**
	 * The sheet button's label.
	 * @property label
	 * @since 0.5.0
	 */
	open var label: Property = Property(string: "")

	/**
	 * The sheet button's style.
	 * @property style
	 * @since 0.5.0
	 */
	open var style: Property = Property(string: "normal")

	/**
	 * The sheet button's image.
	 * @property image
	 * @since 0.5.0
	 */
	open var image: Property = Property()

	//--------------------------------------------------------------------------
	// MARK: JS Properties
	//--------------------------------------------------------------------------

	/**
	 * @method jsGet_label
	 * @since 0.5.0
	 * @hidden
	 */
	@objc open func jsGet_label(callback: JavaScriptGetterCallback) {
		callback.returns(self.label)
	}

	/**
	 * @method jsSet_label
	 * @since 0.5.0
	 * @hidden
	 */
	@objc open func jsSet_label(callback: JavaScriptSetterCallback) {
		self.label = Property(value: callback.value)
	}

	//--------------------------------------------------------------------------

	/**
	 * @method jsGet_style
	 * @since 0.5.0
	 * @hidden
	 */
	@objc open func jsGet_style(callback: JavaScriptGetterCallback) {
		callback.returns(self.style)
	}

	/**
	 * @method jsSet_style
	 * @since 0.5.0
	 * @hidden
	 */
	@objc open func jsSet_style(callback: JavaScriptSetterCallback) {
		self.style = Property(value: callback.value)
	}

	//--------------------------------------------------------------------------

	/**
	 * @method jsGet_image
	 * @since 0.5.0
	 * @hidden
	 */
	@objc open func jsGet_image(callback: JavaScriptGetterCallback) {
		callback.returns(self.image)
	}

	/**
	 * @method jsSet_style
	 * @since 0.5.0
	 * @hidden
	 */
	@objc open func jsSet_image(callback: JavaScriptSetterCallback) {
		self.image = Property(value: callback.value)
	}
}
